syntax = "proto3";

package pspd;

option go_package = "./;pspd";

message User {
  string id = 1;
  string name = 2;
  int32 age = 3;
}

message UserIdRequest {
  string id = 1;
}

message UserResponse {
  User user = 1;
}

message Empty {}

message UsersSummary {
  int32 count = 1;
  repeated string ids = 2;
}

message ScoreRequest {
  string user_id = 1;
  int32 base = 2;
}

message ScoreResponse {
  string user_id = 1;
  double score = 2;
}

message ScoreFragment {
  string user_id = 1;
  double partial = 2;
}

message ChatMessage {
  string user_id = 1;
  string text = 2;
  int64 timestamp = 3;
}

service UserService {
  rpc GetUser (UserIdRequest) returns (UserResponse);                // Unary
  rpc ListUsers (Empty) returns (stream User);                       // Server streaming
  rpc CreateUsers (stream User) returns (UsersSummary);              // Client streaming
  rpc UserChat (stream ChatMessage) returns (stream ChatMessage);    // Bidirectional streaming
}

service StatsService {
  rpc GetScore (ScoreRequest) returns (ScoreResponse);               // Unary
  rpc StreamScores (stream ScoreRequest) returns (stream ScoreResponse); // Bidirectional streaming
}
